src/rdkafka_broker.c

```diff
@@ -351,33 +348,170 @@ void rd_kafka_broker_set_state (rd_kafka_broker_t *rkb, int state) {
 
 
 /**
- * Failure propagation to application.
- * Will tear down connection to broker and trigger a reconnect.
+ * @brief Set, log and propagate broker fail error.
+ *
+ * The caller may omit the format if it thinks this is a recurring
+ * failure, in which case the following things are omitted:
+ *  - log message
+ *  - application OP_ERR
+ *
+ * Dont log anything if this was the termination signal.
+ *
+ * @locks none
+ * @locality broker thread
+ */
+static void rd_kafka_broker_set_error (rd_kafka_broker_t *rkb, int level,
+                                       rd_kafka_resp_err_t err,
+                                       const char *fmt, va_list ap) {
+        char errstr[512];
+        char extra[128];
+        size_t of, ofe;
+        rd_bool_t identical, suppress;
+        int state_duration_ms = (int)((rd_clock() - rkb->rkb_ts_state)/1000);
+
+
+        of = (size_t)rd_vsnprintf(errstr, sizeof(errstr), fmt, ap);
+        if (of > sizeof(errstr))
+                of = sizeof(errstr);
+
+        /* Provide more meaningful error messages in certain cases */
+        if (err == RD_KAFKA_RESP_ERR__TRANSPORT &&
+            !strcmp(errstr, "Disconnected")) {
+                if (rkb->rkb_state == RD_KAFKA_BROKER_STATE_APIVERSION_QUERY) {
+                        /* A disconnect while requesting ApiVersion typically
+                         * means we're connecting to a SSL-listener as
+                         * PLAINTEXT, but may also be caused by connecting to
+                         * a broker that does not support ApiVersion (<0.10). */
+
+                        if (rkb->rkb_proto != RD_KAFKA_PROTO_SSL &&
+                            rkb->rkb_proto != RD_KAFKA_PROTO_SASL_SSL)
+                                rd_kafka_broker_set_error(
+                                        rkb, level, err,
+                                        "Disconnected while requesting "
+                                        "ApiVersion: "
+                                        "might be caused by incorrect "
+                                        "security.protocol configuration or "
+                                        "broker version is < 0.10 "
+                                        "(see api.version.request)",
+                                        ap/*ignored*/);
```

Comment:
Since API requests are handled prior to performing any sort of SASL authentication it may be better to just explicitly state that this is likely a TLS issue. This could be a handshake failure, protocol mismatch etc. Just stating a possible `security.protocol` mismatch may send people chasing SASL issues as they believe TLS to be configured correctly. In my experience people almost always assume its SASL and not TLS when both are in use.

Reply:
Okay, will suggest checking SSL.